### PR TITLE
feat: Add logging to explain reasons for Comet not being able to run a query stage natively

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/385

## Rationale for this change

As a user, I would like to understand why my query stages are not running natively.

## What changes are included in this PR?

New config + logging

## How are these changes tested?

Tested manually:

```
scala> spark.read.parquet("/tmp/test").createOrReplaceTempView("t1")
24/05/07 10:55:16 INFO src/lib.rs: Comet native library initialized
24/05/07 10:55:16 WARN CometSparkSessionExtensions$CometExecRule: Comet cannot execute this plan natively because Execute CreateViewCommand is not supported

scala> spark.sql("select * from t1 where cast(cast(a as double) as decimal(10,2)) > 5").show
24/05/07 10:55:29 WARN CometSparkSessionExtensions$CometExecRule: Comet cannot execute this plan natively because:
	- Comet does not guarantee correct results for cast from DoubleType to DecimalType(10,2) with timezone Some(America/Denver) and evalMode LEGACY (No overflow check). To enable all incompatible casts, set spark.comet.cast.allowIncompatible=true
	- CollectLimit is not supported
+---+
|  a|
+---+
|  9|
|  8|
|  6|
|  7|
+---+
```

### CODE REVIEW

File: common/src/main/scala/org/apache/comet/CometConf.scala
Comment (andygrove) [inline common/src/main/scala/org/apache/comet/CometConf.scala:285]: I'm still thinking about what other options we may want here, or if this should just be a boolean config

**Comment by andygrove** [comment]
@parthchandra How does this look?

File: common/src/main/scala/org/apache/comet/CometConf.scala
Comment (parthchandra) [inline common/src/main/scala/org/apache/comet/CometConf.scala:279]: Do we really need to make this configurable?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:737]: Given that ExecRule is going to be invoked multiple times as planning proceeds, this may get logged multiple times for the same plan. Perhaps that is why we need to keep this configurable.

**Review by parthchandra — APPROVED** [review]
LGTM. Just some observations.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:737]: That is one reason, yes. The logging could get verbose.

Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:737]: I would assume that this feature is useful in development but less so in production since we have the integration with Spark's explain for that.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:738]: Comet could only trigger native execution for partial query plan. This check `isCometNative(newPlan)` only applies for the top operator.

**Comment by andygrove** [comment]
I am moving this to draft for now while I test with TPC-H queries and make some improvements

**Comment by andygrove** [comment]
@viirya @parthchandra This PR is ready for another review. I simplified the logic and removed the check to see if the top level operator is native or not. Instead, if we have recorded any info using `withInfo` then that info will be displayed when the config is enabled. There is an example in the documentation as part of this PR.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:740]: Unrelated, but I feel `ExtendedExplainInfo` can be `object`

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:48]: Perhaps `private[comet]` instead of public?

**Review by kazuyukitanimura — APPROVED** [review]

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:740]: Yes, I had the same thought

Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:740]: It is possible that it has to be a class due to the way this integrates into Spark, but I am not sure.

**Comment by andygrove** [comment]
@viirya I have 2 approvals but need a committer approval to be able to merge.

File: common/src/main/scala/org/apache/comet/CometConf.scala
Comment (viirya) [reply common/src/main/scala/org/apache/comet/CometConf.scala:279]: I think a config should be good. Sometimes it might be verbose.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Looks good to me. Thanks @andygrove

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:740]: Just seeing this. I don't think Spark prevents us from this being an object. I'll take care of this. https://github.com/apache/datafusion-comet/issues/452

**Pull request merged**
